Add unit tests for CategoryService

helenahalldiniths opened the pull request:
This will fix #80

FelixJacobsen committed:
Add .dockerignore file to project

FelixJacobsen committed:
Add files to ignore

helenahalldiniths committed:
Add unit tests for CategoryService

helenahalldiniths committed:
Add unit tests for CategoryService

helenahalldiniths commented:
**Some info:**
- This is more of a draft
- I have commented quite a lot since i'm bad at naming test-methods, feel free to propose other names.
- Since the class mostly is responsible for forwarding requests to the repository the tests feel quite simple, I'm wondering if there is something else that should be tested...?

**Some thoughts I got when working with theses tests (about the rest of the code):**
- The "error handling" of the update method should be moved to the service class.
- The delete method looks quite weird... first we send in the id, then we look for a catogory with said id, then we extract the id from said category and sends it to the "deleteById" method of the repository. Maybe insted we could just call the "deleteById" method and catch the error it throws if the id isn't valid and then throw our costume made exception?

FelixJacobsen committed:
Remove target folder from .dockerignore

helenahalldiniths committed:
Merge pull request #71 from Patlenlix/issue70-.dockerignore

Issue70 .dockerignore

LordRekishi commented:
Will take a look at this once I've gotten a bit further with JMS

LordRekishi committed:
Add dependencies for RabbitMQ and AMQP

LordRekishi committed:
Add Message Queue config file with Beans for queue, exchange, binding, messageConverter (to Json) and rabbit template

LordRekishi committed:
Add CustomMessage entity that will contain the data for the messages we will send

LordRekishi committed:
Add Publisher to publish sample message when using endpoint /publish, with JSON format ("message": "sample message")

LordRekishi committed:
Add Listener to listen for messages from the Publisher and simply print them in the console for demo

LordRekishi committed:
Change LocalDate to java.util.Date to fix bug

LordRekishi committed:
Update README.md with instructions to deploy RabbitMQ message broker, as well as info about login to the management console

LordRekishi committed:
Fix ability to make POST and PUT calls in Insomnia for testing purposes by disabling CSRF (.csrf().disable())

dependabot[bot] committed:
Bump spring-boot-starter-parent from 2.6.4 to 2.6.5

Bumps [spring-boot-starter-parent](https://github.com/spring-projects/spring-boot) from 2.6.4 to 2.6.5.
- [Release notes](https://github.com/spring-projects/spring-boot/releases)
- [Commits](https://github.com/spring-projects/spring-boot/compare/v2.6.4...v2.6.5)

---
updated-dependencies:
- dependency-name: org.springframework.boot:spring-boot-starter-parent
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

LordRekishi committed:
Merge pull request #85 from Patlenlix/dependabot/maven/org.springframework.boot-spring-boot-starter-parent-2.6.5

Bump spring-boot-starter-parent from 2.6.4 to 2.6.5

LordRekishi commented on src/test/java/se/iths/crimedatabase/service/CategoryServiceTest.java:
```diff
@@ -0,0 +1,106 @@
+
+    // CREATE
+    // When we call the create-method with a category,
+    // said category should be passes to the save method of the repository
```
The comments are helpful for reviewing, but I believe we should remove them for the final merge?
Well named test methods are enough, or we could use `@DisplayName("Name here")`

LordRekishi commented on src/test/java/se/iths/crimedatabase/service/CategoryServiceTest.java:
```diff
@@ -0,0 +1,106 @@
+        List<Category> categories = List.of(
+                new Category().setName("Theft").setId(1L),
+                new Category().setName("Burglary").setId(2L),
+                new Category().setName("shoplifting").setId(3L));
```
```suggestion
                new Category().setName("Shoplifting").setId(3L));
```
Let's change to a capital 'S' to correspond with the other categories and look better

LordRekishi commented on src/test/java/se/iths/crimedatabase/service/CategoryServiceTest.java:
```diff
@@ -0,0 +1,106 @@
+    // When we call the findById-method with an id an optional<category> should be returned. It contains a category with that id if it exists
+    // else it's empty.
+    @Test
+    void callingFindByIdShouldReturnAnOptionalWithCorrespondingCategoryIfPresent() {
+        Long validId = 1L;
+        Long invalidId = 2L;
+        Category category = new Category();
+        category.setName("Test").setId(validId);
+        when(repository.findById(validId)).thenReturn(Optional.of(category));
+        when(repository.findById(invalidId)).thenReturn(Optional.empty());
+
+        Optional<Category> expectedCategory = service.findById(validId);
+        Optional<Category> expectedEmpty = service.findById(invalidId);
+
+        assertThat(expectedCategory).isEqualTo(Optional.of(category));
+        assertThat(expectedEmpty).isNotPresent();
+    }
```
Should we perhaps make this into two separate test methods?

LordRekishi commented on src/test/java/se/iths/crimedatabase/service/CategoryServiceTest.java:
```diff
@@ -0,0 +1,106 @@
+    // When we call the delete-method with an id, the category with that id should be deleted if it exists,
+    // else an EntityNotFoundException should be thrown.
+    @Test
+    void callingDeleteWithAValidIdACategoryShouldBeDeletedElseExceptionShouldBeThrown() {
+        Long validId = 1L;
+        Long invalidId = 2L;
+        Category category = new Category().setName("Test").setId(validId);
+        when(repository.findById(invalidId)).thenReturn(Optional.empty());
+        when(repository.findById(validId)).thenReturn(Optional.of(category));
+
+        service.delete(validId);
+
+        verify(repository, times(1)).deleteById(validId);
+        verify(repository, times(0)).deleteById(invalidId);
+        assertThrows(EntityNotFoundException.class, () -> service.delete(invalidId));
+    }
```
Same here? Two methods instead?

LordRekishi reviewed (dismissed):
Looks really nice otherwise, and full coverage! Nice!

FelixJacobsen committed:
Add dependencies for thymeleaf

FelixJacobsen committed:
Add Model-View-Controller config class

FelixJacobsen committed:
Add main.css

FelixJacobsen committed:
Add html templates for all entities

FelixJacobsen committed:
Add release.yml from Storm

FelixJacobsen committed:
Removed commits from other branch

FelixJacobsen committed:
Removed thymeleaf dependency

FelixJacobsen committed:
Update pom.xml

Co-authored-by: Patrik Fallqvist Magnusson <patrik.fallqvist@gmail.com>

LordRekishi committed:
Merge pull request #89 from Patlenlix/Issue24Release

Issue24 release

LordRekishi committed:
Update Dockerfile to fix bug with release

helenahalldiniths committed:
Update src/test/java/se/iths/crimedatabase/service/CategoryServiceTest.java

Co-authored-by: Patrik Fallqvist Magnusson <patrik.fallqvist@gmail.com>

LordRekishi committed:
Revert "Update README.md with instructions to deploy RabbitMQ message broker, as well as info about login to the management console"

This reverts commit 36614c598fc3a533763c9b47bc4a972fd2eb6d68.

helenahalldiniths committed:
Disable csrf for our REST controllers

helenahalldiniths committed:
Merge pull request #96 from Patlenlix/csrf

Disable csrf for our REST controllers

LordRekishi committed:
Add fix

LordRekishi committed:
Add missing files

LordRekishi committed:
Merge remote-tracking branch 'origin/main' into thymeleaf-fix

# Conflicts:
#	src/main/java/se/iths/crimedatabase/view/MvcConfig.java
#	src/main/resources/static/main.css
#	src/main/resources/templates/addresses.html
#	src/main/resources/templates/category.html
#	src/main/resources/templates/crime.html
#	src/main/resources/templates/criminal.html
#	src/main/resources/templates/user.html
#	src/main/resources/templates/victim.html

LordRekishi committed:
Merge remote-tracking branch 'origin/main' into jms-issue-63

# Conflicts:
#	src/main/java/se/iths/crimedatabase/security/SecurityConfig.java

LordRekishi committed:
Update SecurityConfig

LordRekishi committed:
Update CustomMessage

LordRekishi committed:
Update README.md

LordRekishi committed:
Update README.md with instructions and features

LordRekishi committed:
Merge pull request #84 from Patlenlix/jms-issue-63

JMS issue 63 - Basic Version

LordRekishi committed:
Add Insomnia instructions

LordRekishi committed:
Merge pull request #95 from Patlenlix/DockerUpdate

Update Dockerfile to fix bug with release

LordRekishi committed:
Update README.md with Run application instructions (under development)

dependabot[bot] committed:
Bump docker/build-push-action from 2.9.0 to 2.10.0

Bumps [docker/build-push-action](https://github.com/docker/build-push-action) from 2.9.0 to 2.10.0.
- [Release notes](https://github.com/docker/build-push-action/releases)
- [Commits](https://github.com/docker/build-push-action/compare/7f9d37fa544684fb73bfe4835ed7214c255ce02b...ac9327eae2b366085ac7f6a2d02df8aa8ead720a)

---
updated-dependencies:
- dependency-name: docker/build-push-action
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump docker/login-action from 1.12.0 to 1.14.1

Bumps [docker/login-action](https://github.com/docker/login-action) from 1.12.0 to 1.14.1.
- [Release notes](https://github.com/docker/login-action/releases)
- [Commits](https://github.com/docker/login-action/compare/42d299face0c5c43a0487c477f595ac9cf22f1a7...dd4fa0671be5250ee6f50aedf4cb05514abda2c7)

---
updated-dependencies:
- dependency-name: docker/login-action
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

FelixJacobsen committed:
Add thymeleaf dependency

FelixJacobsen committed:
Merge branch 'main' into thymeleaf-fix

FelixJacobsen committed:
Merge pull request #98 from Patlenlix/thymeleaf-fix

Thymeleaf fix

LordRekishi committed:
Merge pull request #100 from Patlenlix/dependabot/github_actions/docker/build-push-action-2.10.0

Bump docker/build-push-action from 2.9.0 to 2.10.0

LordRekishi committed:
Merge pull request #101 from Patlenlix/dependabot/github_actions/docker/login-action-1.14.1

Bump docker/login-action from 1.12.0 to 1.14.1

LordRekishi committed:
Update with notes from meeting 4

LordRekishi committed:
Add docker-compose.yml

LordRekishi committed:
Add configuration for backend and RabbitMQ to work together on the same network

LordRekishi committed:
Update deployment instructions to use docker-compose

FelixJacobsen committed:
Merge pull request #103 from Patlenlix/docker-compose-issue-53

Docker compose issue 53

FelixJacobsen committed:
Merge pull request #102 from Patlenlix/Update-Notes-meeting-4

Update with notes from meeting 4

LordRekishi committed:
Update Deployment instructions

FelixJacobsen committed:
Update with release tag

dependabot[bot] committed:
Bump spring-boot-starter-thymeleaf from 2.6.4 to 2.6.5

Bumps [spring-boot-starter-thymeleaf](https://github.com/spring-projects/spring-boot) from 2.6.4 to 2.6.5.
- [Release notes](https://github.com/spring-projects/spring-boot/releases)
- [Commits](https://github.com/spring-projects/spring-boot/compare/v2.6.4...v2.6.5)

---
updated-dependencies:
- dependency-name: org.springframework.boot:spring-boot-starter-thymeleaf
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

LordRekishi committed:
Merge pull request #109 from Patlenlix/dependabot/maven/org.springframework.boot-spring-boot-starter-thymeleaf-2.6.5

Bump spring-boot-starter-thymeleaf from 2.6.4 to 2.6.5

helenahalldiniths committed:
Refactor code and remove comments

helenahalldiniths committed:
Merge remote-tracking branch 'origin/issues80' into issues80

# Conflicts:
#	src/test/java/se/iths/crimedatabase/service/CategoryServiceTest.java

helenahalldiniths commented:
I have updated the branch and made some small changes.

It looks like I have made changes to 22 files... This is because I have rebased the new commits onto this branch. If you checkout the code and compare it with the current main branch you can see that the only difference is the test class.

LordRekishi commented:
I think we should try to use merge more than rebase. Merge main into our branch before pushing. That will solve all these commits doubling up.
